relay/network_graph/alg.py

```diff
@@ -0,0 +1,129 @@
+"""graph algorithms"""
+
+import heapq
+from typing import List, Set, Dict, Callable
+import networkx as nx
+import abc
+
+
+class CostAccumulator(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def zero(self):
+        """return 'zero cost' element, which is the initial cost for a one-node path
+
+        This must be less than or equal to any value
+        accumulate_cost_for_next_hop returns.
+        """
+        pass
+
+    @abc.abstractmethod
+    def accumulate_cost_for_next_hop(
+        self, cost_from_start_to_node, node, dst, graph_data
+    ):
+        """
+        compute the total cost from one of the starting nodes to dst via node
+        given that the cost from one of the starting nodes to node is
+        cost_from_start_to_node.
+
+        In other words this adds the cost for the hop from node to dst to
+        cost_from_start_to_node
+
+        This function must return a value greater or equal to the given
+        cost_from_start_to_node parameter.
```

Comment:
We should also hint in the documentation that this method can return null to signal that this is a forbidden path.